### ISSUE TITLE
support Docker HEALTHCHECK

### ISSUE THREAD

**Issue by bradrydzewski** [drone/drone#2087]
support a healthcheck for the server and agent

```
HEALTHCHECK CMD curl --fail http://localhost:3000/ || exit 1
```

**Comment by gtaylor** [comment]
This will be great for Kubernetes livelieness/readiness checks.

**Comment by dellintosh** [comment]
Also Mesos... ;)  👍

**Comment by tboerger** [comment]
And even plain docker-compose ;)

**Comment by mrueg** [comment]
Agents could probably use https://github.com/grpc/grpc/blob/master/doc/health-checking.md and report it via a local debug socket.

**Comment by omerxx** [comment]
Hey everyone, any chance this gets priority anytime soon?
I'm trying to deploy 0.8 on AWS ECS.

**Comment by vanadium23** [comment]
@bradrydzewski @gtaylor it's just adding this line to Dockerfile? Can I take it?
Also may be we need separate endpoint for healtcheck as described in Mozilla [Dockerflow](https://github.com/mozilla-services/Dockerflow).

**Comment by bradrydzewski** [comment]
The first step is to propose a design. The server image does not have curl or any utilities available to ping the user interface (by design), which means a command needs to be added to the drone-server binary. For example `drone-server --ping`

The agent image does not have any ports that we can ping for healthcheck data. So the agent needs to be modified to add some sort of hooks for healthchecks. Similar to the server, we then need a command to ping the agent.

So the first step is to document a design. Once the design is approved we can begin coding, and yes, @vanadium23 would love to have your help coding this up.

**Comment by tboerger** [comment]
I'm also really interested in that as drone is one of only a few services for me without a healthcheck. What should the agent --ping or --healthcheck flag check for the agent? If a process on the system (container) exists since we are not opening any port?

**Comment by tboerger** [comment]
Or should we just check if the agent is able to connect to the server?

**Comment by bradrydzewski** [comment]
> Or should we just check if the agent is able to connect to the server?

I propose initial implementation agent healthcheck should just return a 200 response if the agent is running. Future versions of the agent should absolutely perform internal checks and return 500 if there are issues pinging the server or blocked builds, but I would like to keep that out of scope for the initial implementation since the agent grpc code is new and still evolving.

**Comment by gtaylor** [comment]
:+1: A naive 200 would at least be be immediately helpful for detecting complete deadlock or other weird states.

**Comment by metalmatze** [comment]
That naive health check can also happen on some normal http page. :relaxed:

**Comment by bradrydzewski** [comment]
I started working on an initial healthcheck for the agent, however, I am not sure that a simple endpoint that returns a 200 will be enough here. Even if the agent is deadlocked or no longer processing builds, the http endpoint will likely still serve requests. If the agent dies, the docker container will exit with a non-zero exit code, which I presume kubenerets will detect, and then restart the image (based on whatever policy you have configured).

So I think to make the healthcheck meaningful we need to come up with a list of failure points and how we can detect these failure points. This information can then be used with the healthcheck endpoint.

Here are some things that can go wrong, off the top of my head:

1. agent cannot connect to server, will perform infinite backoff to reconnect
2. agent cannot make grpc call (usually due to server connection), will perform infinite retries
3. docker daemon locks
4. some internal lock / race / bug that we don't know about yet

Number 3 and 4 could in theory be detected by keeping track of how long the build has been running. If the runtime has exceeded the timeout + buffer and the agent is still processing the build, we know there is probably an issue ...

**Comment by gtaylor** [comment]
I like the conditions you've outlined. As an side, having a stupid 200 healthcheck would still be useful for Readiness probes, since we could at least see that the process is in good enough shape to return the 200s (rather than deadlocked entirely or in some other weird state).

**Comment by bradrydzewski** [comment]
sure thing. can I get a quick review https://github.com/drone/drone/pull/2209 ?

**Comment by bradrydzewski** [comment]
ok, merged #2209 with the more advanced logic mentioned above. This will be included in 0.8.0-rc.5 and should therefore get bundled with 0.8.0-final pending we don't discover any last minute issues.

note that this change also gives you the option to expose port 3000 for the agent to access the following endpoint:

* `/varz` returns some basic agent statistics, including currently running builds
* `/healthz` returns 200 if agent is healthy
* `/version` returns returns the version information
